packages/gatsby-plugin-mdx/src/gatsby-node.ts

```diff
@@ -1,53 +1,90 @@
-import type { Options } from "@mdx-js/loader"
+import type { IFileNode, NodeMap } from "./types"
 
 import path from "path"
 import { sentenceCase } from "change-case"
 import fs from "fs-extra"
 import grayMatter from "gray-matter"
 
-import { defaultOptions, IMdxPluginOptions } from "./plugin-options"
-import { IMdxLoaderOptions } from "./mdx-loader"
+import {
+  defaultOptions,
+  enhanceMdxOptions,
+  IMdxPluginOptions,
+} from "./plugin-options"
+import { IGatsbyLayoutLoaderOptions } from "./gatsby-layout-loader"
+import compileMDX from "./compile-mdx"
+import { IGatsbyMDXLoaderOptions } from "./gatsby-mdx-loader"
 
 /**
  * Add support for MDX files including using Gatsby layout components
  */
-export const onCreateWebpackConfig: GatsbyNode["onCreateWebpackConfig"] = (
-  { actions, loaders, getNodesByType },
-  pluginOptions
-) => {
-  const fileNodes = getNodesByType(`File`)
-  const fileMap = new Map()
-  fileNodes.forEach(fileNode => fileMap.set(fileNode.absolutePath, fileNode))
-
-  actions.setWebpackConfig({
-    module: {
-      rules: [
-        {
-          test: /\.mdx?$/,
-          use: [
-            loaders.js(),
-            {
-              loader: `@mdx-js/loader`,
-              options: {
-                useDynamicImport: true,
-                providerImportSource: `@mdx-js/react`,
-              } as Options,
-            },
-            {
-              loader: path.join(`gatsby-plugin-mdx`, `dist`, `mdx-loader`),
-              options: {
-                fileMap,
-                pluginOptions,
-              } as IMdxLoaderOptions,
-            },
-          ],
-        },
-      ],
-    },
-  })
-}
+export const onCreateWebpackConfig: GatsbyNode["onCreateWebpackConfig"] =
+  async (
+    { actions, loaders, getNode, getNodesByType, pathPrefix, reporter, cache },
+    pluginOptions
+  ) => {
+    const mdxNodes = getNodesByType(`Mdx`)
+    const nodeMap: NodeMap = new Map()
+    mdxNodes.forEach(mdxNode => {
+      if (!mdxNode.parent) {
+        return
+      }
+      const fileNode = getNode(mdxNode.parent) as IFileNode
+      if (!fileNode) {
+        return
+      }
+      nodeMap.set(fileNode.absolutePath, { fileNode, mdxNode })
+    })
+
+    const options = defaultOptions(pluginOptions as IMdxPluginOptions)
+
+    const mdxOptions = await enhanceMdxOptions(
+      pluginOptions as IMdxPluginOptions,
+      {
+        getNode,
+        getNodesByType,
+        pathPrefix,
+        reporter,
+        cache,
+      }
+    )
+
+    actions.setWebpackConfig({
+      module: {
+        rules: [
+          {
+            test: /\.mdx?$/,
+            use: [
+              loaders.js(),
+              {
+                loader: path.join(
+                  `gatsby-plugin-mdx`,
+                  `dist`,
+                  `gatsby-mdx-loader`
+                ),
+                options: {
+                  options: mdxOptions,
+                  nodeMap,
+                } as IGatsbyMDXLoaderOptions,
```

Comment:
Same as above, you'll want to type `mdxOptions` and `nodeMap` in a way that both will have the types like `IGatsbyMDXLoaderOptions` instead of using type assertions